testing-mainnet protocols (#26)
studydefi/studydefi.com: content/blog/how-to-integrate-with-mainnet-protocols/index.md

---

> title: "How to Integrate with Ethereum Mainnet DeFi Protocols"

Reviewer: Pick one of the suggestions below:

```suggestion
title: "Developing on Ethereum's existing Mainnet DeFi protocols"
```

```suggestion
title: "Integrating Mainnet DeFi Protocols into your Ethereum Dapp"
```

```suggestion
title: "Working off a forked instance of the Ethereum Mainnet"
```

```suggestion
title: "How to develop on a forked sandbox of the Ethereum Mainnet"
```

date: "2020-04-24"
description: "Wanna try build a dapp that interacts with multiple mainnet protocols but don't know where to start? Read on!"
---

While building out [Dedge](https://dedge.exchange), one of the first problems we encountered was:

> > How and where do we start integrating with mainnet DeFi protocols?

Reviewer: ```suggestion
> How do we even begin to start interacting with existing Mainnet dapps?
```

> Ideally, we would like to fork off the current mainnet state into a sandboxed environment, and develop/test our smart contracts from there. The emphasis on forking off the mainnet state it provides a couple key features:

Reviewer: ```suggestion
Ideally, we would like to fork off the current Mainnet state into a sandboxed environment, and develop/test our smart contracts from there. The reason this is important is because it provides a few essential benefits:
```

> 1. Liquidity automatically configured on other DeFi procotols e.g. Aave/Compound/Compound.

Reviewer: ```suggestion
1. All existing DeFi protocols (e.g. Aave/Compound/etc.) have real and full liquidity.
```

> 2. You're almost guaranteed every major procotol will be deployed on mainnet.

Reviewer: ```suggestion
2. You have guaranteed access to every major protocol deployed on Mainnet.
```

>     - If you're interacting with multiple protocols, they might not all be available on the same testnet e.g. Rinkeby

Reviewer: ```suggestion
    - If you were using official testnets (e.g. Rinkeby) and needed to interact with more than one protocol, not all of them might be deployed to the same network
```

> 3. As you're forking off the mainnet state, all mainnet contract addresses will be the same in your private testnet.

Reviewer: ```suggestion
3. Since you're forking off of Mainnet state, all contract addresses will be the same in both development and production.
```

> 4. Increases your development velocity as you can just focus on developing your application, not writing scripts to deploy the protocols to your private testnet.

Reviewer: ```suggestion
4. Your development velocity is significantly increased because you don't have to worry about deploying existing protocols onto your test network.
```

> Fortunately, the amazing folks over at [truffle](https://www.trufflesuite.com/) has made a lovely tool called [ganache-cli](https://www.trufflesuite.com/ganache).

Reviewer: ```suggestion
Fortunately, the amazing folks over at [truffle](https://www.trufflesuite.com/) has made a lovely tool called [Ganache](https://github.com/trufflesuite/ganache-cli/).
```

# Quickstart

> To fork off mainnet, simply invoke the `-f` flag on `ganache-cli`.

Reviewer: ```suggestion
To fork off Mainnet, simply invoke the `-f` flag on `ganache-cli`.
```

```bash
npx ganache-cli -f https://mainnet.infura.io/v3/7d0d81d0919f4f05b9ab6634be01ee73 -i 5777 -d
```

> _The infura api key is obtained from [ethers.js](https://github.com/ethers-io/ethers.js/blob/master/src.ts/providers/infura-provider.ts#L14)._

Reviewer: ```suggestion
_This Infura API key is obtained publicly from [Ethers.js](https://github.com/ethers-io/ethers.js/blob/master/src.ts/providers/infura-provider.ts#L14)._
```

> You should be able to interact with ganache (with mainnet state) on `http://localhost:8454`.

Reviewer: ```suggestion
You should be able to interact with Ganache (with Mainnet state) on `http://localhost:8545`.
```

> You can then try and interact with some mainnet contract, for example `wETH`:

Reviewer: ```suggestion
You can then try and interact with some Mainnet contracts, for example `WETH` or Wrapped Ether:
```

```javascript
const { legos } = require("@studydefi/money-legos");
const { ethers } = require("ethers");

const provider = new ethers.providers.JsonRpcProvider("http://localhost:8545");

const wallet = new ethers.Wallet(
  "0xb0057716d5917badaf911b193b12b910811c1497b5bada8d7711f758981c3773", // Default private key for ganache-cli -d
  provider
);

const wethContract = new ethers.Contract(
  legos.erc20.weth.address,
  legos.erc20.weth.abi,
  wallet
);

const main = async () => {
  await wethContract.deposit({
    value: ethers.utils.parseEther("1.0"),
    gasLimit: 1000000,
  });

  const wethBal = await wethContract.balanceOf(wallet.address);

>   console.log(`wETH Bal: ${ethers.utils.formatEther(wethBal)}`);

Reviewer: ```suggestion
  console.log(`WETH Balance: ${ethers.utils.formatEther(wethBal)}`);
```

};

main();
```

# "Gotchas" to be aware of

> Unfortunately the infura key from `ether.js` does not point to an archival node. That means you will have to to restart `ganache-cli` every 128 blocks (~30 minutes), otherwise you'll be greeted with the following message from infura:

Reviewer: ```suggestion
Unfortunately the infura key from Ethers.js does not point to an archival node. That means you will have to to restart `ganache-cli` every 128 blocks (~30 minutes), otherwise you'll be greeted with the following message from Infura:
```

> Returned error: project ID does not have access to archive state

> Stay tune for the next blogpost on how to avoid this issue (without paying $250/month for an archival node)!

Reviewer: ```suggestion
Stay tuned for the next blogpost on how to avoid this issue (without paying $250/month for an archival node)!
```
